src/java.base/share/classes/java/text/ListFormat.java

```diff
@@ -0,0 +1,571 @@
+        // get pattern strings
+        var m = PARSE_START.matcher(patterns[START]);
+        String startBefore;
+        String startBetween;
+        if (m.matches()) {
+            startBefore = m.group(1);
+            startBetween = m.group(2);
+        } else {
+            throw new IllegalArgumentException("start pattern is incorrect: " + patterns[START]);
+        }
+        m = PARSE_MIDDLE.matcher(patterns[MIDDLE]);
+        if (m.matches()) {
+            middleBetween = m.group(1);
+        } else {
+            throw new IllegalArgumentException("middle pattern is incorrect: " + patterns[MIDDLE]);
+        }
+        m = PARSE_END.matcher(patterns[END]);
+        String endBetween;
+        String endAfter;
+        if (m.matches()) {
+            endBetween = m.group(1);
+            endAfter = m.group(2);
+        } else {
+            throw new IllegalArgumentException("end pattern is incorrect: " + patterns[END]);
+        }
+
+        // generate two/three patterns, if empty
+        if (patterns[TWO] == null || patterns[TWO].isEmpty()) {
+            patterns[TWO] = startBefore + "{0}" + endBetween + "{1}" + endAfter;
+        }
+        if (patterns[THREE] == null || patterns[THREE].isEmpty()) {
+            patterns[THREE] = startBefore + "{0}" + startBetween + "{1}" + endBetween + "{2}" + endAfter;
+        }
+        startPattern = Pattern.compile(startBefore + "(.+?)" + startBetween);
+        endPattern = Pattern.compile(endBetween + "(.+?)" + endAfter);
+    }
+
+    /**
+     * {@return the available locales that support list formatting}
+     */
+    public static Locale[] getAvailableLocales() {
+        // Same as a typical format class
+        return DateFormat.getAvailableLocales();
+    }
+
+    /**
+     * {@return the list format object for the default
+     * {@link Locale.Category#FORMAT FORMAT Locale}, {@code STANDARD} type,
+     * and {@code FULL} style}
+     */
+    public static ListFormat getInstance() {
+        return getInstance(Locale.getDefault(Locale.Category.FORMAT), Type.STANDARD, Style.FULL);
+    }
+
+    /**
+     * {@return the list format object for the specified {@code Locale}, {@code Type},
+     * and {@code Style}}
+     * @param locale Locale to be used, not null
+     * @param type type of the list format. One of STANDARD/OR/UNIT, not null
+     * @param style style of the list format. One of FULL/SHORT/NARROW, not null
+     * @throws NullPointerException if any of the arguments are null
+     */
+    public static ListFormat getInstance(Locale locale, Type type, Style style) {
+        Objects.requireNonNull(locale);
+        Objects.requireNonNull(type);
+        Objects.requireNonNull(style);
+        return new ListFormat(locale, LocaleProviderAdapter.forType(LocaleProviderAdapter.Type.CLDR)
+                .getLocaleResources(locale)
+                .getListPatterns(type, style));
+    }
+
+    /**
+     * {@return the list format for the specified patterns}
+     * <p>
```

Comment:
LDML includes a section on the "Gender of Lists";  If later support was desired, is this API extensible to include that?  https://www.unicode.org/reports/tr35/tr35-general.html#List_Gender

Reply:
IIUC, although it is described within "list patterns" in LDML, it is orthogonal to "list patterns." The purpose of the "Gender of Lists" seems to me the resulting gender of the whole list of names, e.g., "Tom and Mary" is "other" in English, but it is "male" in Arabic (I just used the sample XML in the spec). So I believe it is not an extension of the class.